cmd/object-handlers.go

```diff
@@ -2808,20 +2810,52 @@ func (api objectAPIHandlers) GetObjectTaggingHandler(w http.ResponseWriter, r *h
 		return
 	}
 
-	// Allow getObjectTagging if policy action is set.
-	if s3Error := checkRequestAuthType(ctx, r, policy.GetObjectTaggingAction, bucket, object); s3Error != ErrNone {
-		writeErrorResponse(ctx, w, errorCodes.ToAPIErr(s3Error), r.URL)
-		return
-	}
-
 	opts, err := getOpts(ctx, r, bucket, object)
 	if err != nil {
 		writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL)
 		return
 	}
 
-	// Get object tags
-	ot, err := objAPI.GetObjectTags(ctx, bucket, object, opts)
+	getObjectInfo := objAPI.GetObjectInfo
+	if api.CacheAPI() != nil {
+		getObjectInfo = api.CacheAPI().GetObjectInfo
+	}
+
+	var ot *tags.Tags
+	// Allow getObjectTagging if policy action is set.
+	if s3Error := checkRequestAuthType(ctx, r, policy.GetObjectTaggingAction, bucket, object); s3Error != ErrNone {
+		if getRequestAuthType(r) != authTypeAnonymous { // For authenticated requests if policy returns error, return error right away.
+			writeErrorResponse(ctx, w, errorCodes.ToAPIErr(s3Error), r.URL)
+			return
+		} // for anonymous check for metadata
+
+		objInfo, gerr := getObjectInfo(ctx, bucket, object, opts)
+		if err != nil {
+			writeErrorResponse(ctx, w, toAPIError(ctx, gerr), r.URL)
+			return
+		}
+
+		if objInfo.UserTags != "" {
+			// Set this such that authorization policies can be applied on the object tags.
+			r.Header.Set(xhttp.AmzObjectTagging, objInfo.UserTags)
+		}
+
+		if !globalPolicySys.IsAllowed(policy.Args{
+			Action:          policy.GetObjectTaggingAction,
+			BucketName:      bucket,
+			ObjectName:      object,
+			ConditionValues: getConditionValues(r, "", auth.AnonymousCredentials),
+			IsOwner:         false,
+		}) {
+			writeErrorResponse(ctx, w, errorCodes.ToAPIErr(s3Error), r.URL)
+			return
+		}
+
+		ot, err = tags.ParseObjectTags(objInfo.UserTags)
+	} else {
+		// Get object tags
+		ot, err = objAPI.GetObjectTags(ctx, bucket, object, opts)
+	}
```

Comment:
I am not sure I understand this else block, it looks like we still continue if there is any error in authenticating the request